src/clientlayers/RetryRequest.jl

```diff
@@ -77,6 +77,9 @@ function retrylayer(handler)
 end
 
 isrecoverable(ex) = true
+isrecoverable(ex::ArgumentError) = false
```

Comment:
is `isrecoverable(ex) = true` just the wrong default?

e.g. what if CloudBase stops throwing a `ArgumentError` and decides to throw a `MalformedTokenError`?

Reply:
>can you explain why this is the right solution?

This seemed like the least disruptive fix for the problem we've seen. Maybe the right solution is to formalize the contract a `layer` needs to fulfill wrt throwing exceptions.

>is isrecoverable(ex) = true just the wrong default?

I think Jacob planned a bigger refactor for retries, switching exception retries from opt-out to opt-in would make sense to me.

> what if CloudBase stops throwing a ArgumentError and decides to throw a MalformedTokenError?

I understand, but I'm less worried about this possibility since this is the first instance of this issue we've seen and I don't expect the underlying packages to make such changes without us noticing.

Reply:
i think this bug (i.e. we'd retry on exceptions happening inside layers) shows that `isrecoverable(ex) = true` is the wrong default. What would be the disruption of flipping that to `isrecoverable(ex) = false`? 

> I don't expect the underlying packages to make such changes without us noticing

packages like CloudBase.jl are _users_ on HTTP.jl (not dependencies), and we don't have much visibility on users, so i am worried about other users running into this same issue (just instead of via an ArgumentError via a MethodError or a AssertionError or a BoundError or ...)

Reply:
If @quinnj is not around, perhaps we can get a third view from @NHDaly (and/or whoever was running into this issue)?

Reply:
Looks like Blame shows this as the PR that set that as the default: https://github.com/JuliaWeb/HTTP.jl/pull/990

Looks like maybe the PR was trying to change to _always_ retry if the request is retryable (has a retryable body or is idempotent)? But yeah... it does seem like a strange default...

Do you have any memory of why that PR was made, @nickrobinson251?

Reply:
i think that PR was a fix for some oversight in the preceding refactor: https://github.com/JuliaWeb/HTTP.jl/pull/974/files

I don't believe that PR intended to change any behaviour, only increase flexibility. The description (from the previous draft of the same thing https://github.com/JuliaWeb/HTTP.jl/pull/877) was
> this PR proposes **more configurability for retry logic** in the hopes that users won't opt to roll their own incorrect retry logic.

and i notice that before that refactor [we had
](https://github.com/JuliaWeb/HTTP.jl/pull/974/files#diff-b2c5d0a3fdbbe9ee74c7d758e6e600a881b844302b7dfe83a0d4a944caebfc81L66)
```julia
isrecoverable(e) = false 
isrecoverable(e::Union{Base.EOFError, Base.IOError, MbedTLS.MbedException, OpenSSL.OpenSSLError}) = true
isrecoverable(e::ArgumentError) = e.msg == "stream is closed or unusable"
isrecoverable(e::Sockets.DNSError) = true
isrecoverable(e::ConnectError) = true
isrecoverable(e::RequestError) = isrecoverable(e.error)
isrecoverable(e::StatusError) = retryable(e.status)
```
which makes me think the default of `true` was unintended

Reply:
Amusingly, there is now a failing test:

```julia
@test HTTP.RetryRequest.isrecoverable(ErrorException(""))
```

so my attempt to make an improvement with as little behavior change as possible was doomed to fail from the beginning.

Reply:
>What would be the disruption of flipping that to isrecoverable(ex) = false?

The fact that we wouldn't retry because we relied on certain exceptions being considered recoverable implicitly. Even this current PR seem to have missed `isrecoverable(e::ArgumentError) = e.msg == "stream is closed or unusable"` which is unfortunate.